File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        
        vector<vector<int>> ans;
        
        int start = newInterval[0];
        int end = newInterval[1];
        int n = intervals.size();
        
        int i = 0;

        //push back all the intervals which are not overlapping
        while(i < n && intervals[i][1] < start){
            
            ans.push_back(intervals[i]);
            i++;
            
        } 

        //merge the intervals which are overlapping
        vector<int> mergedInterval = {start, end};
        
        while(i < n && intervals[i][0] <= end){
            
            mergedInterval[0] = min(mergedInterval[0], intervals[i][0]);
            mergedInterval[1] = max(mergedInterval[1], intervals[i][1]);
            
            i++;
            
        }
        
        ans.push_back(mergedInterval);

        //push back all ther remaining intervals which aren't overlapping
        
        while(i < n){
            
            ans.push_back(intervals[i]);
            i++;
        
        }

        return ans;        
    }
};

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
        
   public int[][] insert(int[][] intervals, int[] ni) {
       
        if (intervals.length == 0) {
            return new int[][]{ni};
        }
        
        int m = ni[0];
        int l = ni[1];
        
        for (int[] interval : intervals) {
            
            if (m <= interval[1] && l >= interval[0]) {
                m = Math.min(interval[0], m);
                l = Math.max(interval[1], l);
            }
            
        }
        
        List<List<Integer>> ans = new ArrayList<>();
        boolean f = false;
        
        for (int[] interval : intervals) {
            if (m <= interval[1] && l >= interval[0]) {
                if (!f) {
                    f = true;
                    List<Integer> temp = new ArrayList<>();
                    temp.add(m);
                    temp.add(l);
                    ans.add(temp);
                }
                continue;
            }
            ans.add(Arrays.asList(interval[0], interval[1]));
        }
        
        if (!f) {
            ans.add(Arrays.asList(m, l));
        }

        // Convert List of Lists to a 2D array
        int[][] result = new int[ans.size()][];
       
        for (int i = 0; i < ans.size(); i++) {
            
            List<Integer> row = ans.get(i);
            result[i] = new int[row.size()];
            
            for (int j = 0; j < row.size(); j++) {
                result[i][j] = row.get(j);
            }
            
        }

        if (!f) {
            Arrays.sort(result, Comparator.comparingInt(a -> a[0]));
        }
        return result;
    }    
}